message_definitions/v1.0/common.xml

```diff
@@ -3150,7 +3150,7 @@
       <field type="uint16_t" name="param_index">Index of this onboard parameter</field>
     </message>
     <message id="23" name="PARAM_SET">
-      <description>Set a parameter value (write new value to permanent storage). IMPORTANT: The receiving component should acknowledge the new parameter value by sending a PARAM_VALUE message to all communication partners. This will also ensure that multiple GCS all have an up-to-date list of all parameters. If the sending GCS did not receive a PARAM_VALUE message within its timeout time, it should re-send the PARAM_SET message. The parameter microservice is documented at https://mavlink.io/en/services/parameter.html</description>
+      <description>Set a parameter value (write new value to permanent storage). Within a transaction the recieving componenent should respond with PARAM_ACK_TRANSACTION to the setter component. IMPORTANT: If sent outside a transaction the receiving component should acknowledge the new parameter value by broadcasting a PARAM_VALUE message to all communication partners (broadcasting ensures that multiple GCS all have an up-to-date list of all parameters). If the sending GCS did not receive a PARAM_VALUE or PARAM_ACK_TRANSACTION message within its timeout time, it should re-send the PARAM_SET message. The parameter microservice is documented at https://mavlink.io/en/services/parameter.html</description>
```

Comment:
`PARAM_ACK_TRANSACTION`?

Interesting idea, but we don't do that.  And it would require some discussion to determine if it is a good idea....

We should probably have a spell-checker running over these text messages in
 autotest...

```suggestion
      <description>Set a parameter value (write new value to permanent storage). Within a transaction the receiving component should respond with PARAM_ACK_TRANSACTION to the setter component. IMPORTANT: If sent outside a transaction the receiving component should acknowledge the new parameter value by broadcasting a PARAM_VALUE message to all communication partners (broadcasting ensures that multiple GCS all have an up-to-date list of all parameters). If the sending GCS did not receive a PARAM_VALUE or PARAM_ACK_TRANSACTION message within its timeout time, it should re-send the PARAM_SET message. The parameter microservice is documented at https://mavlink.io/en/services/parameter.html</description>
```

@hamishwillee

Reply:
@peterbarker Discussion happened in the dev call. It is a wrapper/independent of the service that is made transactional, so I don't think we considered in that discussion that docs on this might need to be updated. FWIW I think that needs to be made clear - ie transactions need to be documented and in this case the transaction case should perhaps come after the "normal case" - but I do think this should be here.

@auturgy  You were there during the discussion - thoughts on best way to handle this one. Either way, @julianoes can you remember if you were tasked to document this?

Reply:
@hamishwillee that would be @DanielePettenuzzo, I think.

Reply:
@DaniellePettenuzzo can you please comment on this???

Reply:
OK, so just FYI, this was merged in https://github.com/mavlink/mavlink/pull/1428 and there was discussion in the dev call in the topic on [20200722](https://github.com/mavlink/mavlink/wiki/20200722-Dev-Meeting#notes). Those notes suggested that the extended param protocol would also be updated but that was un-done (not worth the effort for the bytes saved).

In summary, the use case is allowing dependent "things" to be set together - PX4 uses this for wifi settings that make no sense independently.

@peterbarker This was discussed in dev call with @auturgy present. Could you please arrange a review of the concept with ArduPilot if merging this is a problem? 

If it would help, I'd quite like to re-write this to make it clear that transactions are separate from parameters and that it is perfectly reasonable not to implement transaction protocol. So something like:

> Set a parameter value (write new value to permanent storage). The receiving component should acknowledge the new parameter value by broadcasting a PARAM_VALUE message to all communication partners (broadcasting ensures that multiple GCS all have an up-to-date list of all parameters). If the sending GCS did not receive a PARAM_VALUE within its timeout time, it should re-send the PARAM_SET message. The parameter microservice is documented at https://mavlink.io/en/services/parameter.html
> 
> PARAM_SET may also be called within the context of a transaction: [link to docs]. Within a transaction the receiving component should respond with PARAM_ACK_TRANSACTION to the setter component (instead of PARAM_VALUE), and PARAM_SET should be re-sent if this is not received.

Reply:
That looks like a reasonable rewrite.  I'd suggest you could drop off "to all communication partners" as that implies you know of all that might receive that broadcast.

I'm tempted to suggest ArduPilot leave off the last paragraph entirely (its nice it is broken out) so the diff is minimal.

AFAICS there's no implementation of parameter transactions within PX4 Firmware.....

.... also, how do you know you have received all parameter-sets relevant to the current transaction?  AFAICS you could receive a param-set-ack from a previous transaction and be happy its been received from your current transaction... seems like the opaque transaction ID should've been included in the acks.

Reply:
OK, I've updated in https://github.com/mavlink/mavlink/pull/1545 - look OK to you?

> I'm tempted to suggest ArduPilot leave off the last paragraph entirely (its nice it is broken out) so the diff is minimal.
  
FWIW I'd rather you left it in. If this is useful, let's make it as common as possible.

> AFAICS there's no implementation of parameter transactions within PX4 Firmware.....

Yet. Auterion use it - AFAIK just late to push upstream. I will query internally.

> .... also, how do you know you have received all parameter-sets relevant to the current transaction? AFAICS you could receive a param-set-ack from a previous transaction and be happy its been received from your current transaction... seems like the opaque transaction ID should've been included in the acks.

There was some such discussion. I think the decision was that it wasn't needed because the mavlink state machines all only handle commands in sequence. So you shouldn't be able to get an ack from the old sequence.